python/letsql/expr/relations.py

```diff
@@ -236,3 +163,72 @@ def make_unbound_dt(self):
             name=name,
             schema=self.schema,
         )
+
+
+def register_and_transform_remote_tables(expr):
+    import letsql as ls
+
+    tables = {}
+    created = {}
+    seen_expr = {}
+
+    def get_batches(ex):
+        if ex not in seen_expr:
+            if not ex.op().find((RemoteTable, MarkedRemoteTable, CachedNode, Read)):
+                batches = ex.to_pyarrow_batches()  # execute in native backend
+            else:
+                batches = ls.to_pyarrow_batches(ex)
+        else:
+            batches = seen_expr[ex]
+
+        schema = ex.as_table().schema().to_pyarrow()
+        result, keep = SafeTee.tee(batches, 2)
+        seen_expr[ex] = keep
+        return pa.RecordBatchReader.from_batches(schema, result)
+
+    def replacer(node, _, **kwargs):
+        if isinstance(node, Relation):
+            updated = {}
+            for k, v in list(kwargs.items()):
+                try:
+                    if v in tables:
+                        remote: RemoteTableCounter = tables[v]
+                        name = f"{v.name}_{next(remote.count)}"
+                        kwargs[k] = MarkedRemoteTable(
+                            name,
+                            schema=v.schema,
+                            source=v.source,
+                            namespace=v.namespace,
+                            remote_expr=v.remote_expr,
+                        )
+
+                        batches = get_batches(remote.remote_expr)
```

Comment:
I think we should call `get_batches` only once

Reply:
We need to register one batch per kwarg, each kwarg representing a reference to the RemoteTable. It is similar to searching for occurrences in the SQL string.

Reply:
I'm saying we should invoke the tee'ing only once: we should not tee a tee.

Reply:
Done